_posts/2025-04-28-sparse-autodiff.md

```diff
@@ -699,13 +697,16 @@ The complexity of a single HVP scales roughly with the complexity of the functio
 The Hessian has a **symmetric** structure (equal to its transpose), which means that matrix-vector products and vector-matrix products coincide.
 This specificity can be exploited in the sparsity detection as well as in the coloring phase.
 
-### Pattern detection
+### Second order pattern detection
 
 Detecting the sparsity pattern of the Hessian is more complicated than for the Jacobian.
-This is because, in addition to the usual linear dependencies, we now have to account for **nonlinear interactions** between every pair of coefficients.
-For instance, if $f(\mathbf{x})$ involves a term of the form $x_1 + x_2$, it will not affect the Hessian. 
-On the other hand, a term $x_1 x_2$ will yield two equal non-zero coefficients, one at position $(1, 2)$ and one at position $(2, 1)$.
-Thus, the abstract interpretation system used for detection needs a finer classification of operators, to distinguish between locally linear ones (sum, max) and locally nonlinear ones (product, exp).
+This is because, in addition to the usual linear dependencies, we now have to account for **nonlinear interactions** in the compute graph.
+
+For instance, if $f(\mathbf{x})$ involves a term of the form $x_1 + x_2$, it will not directly affect the Hessian. 
+On the other hand, multiplying this term with $x_3$ to obtain an output $f(\mathbf{x}) = (x_1 + x_2)\,x_3$ 
+will yield non-zero coefficients at position $(1, 3)$, $(3, 1)$, $(2, 3)$ and $(3, 2)$.
+Thus, the abstract interpretation system used for detection needs a finer classification of operators, 
+distinguishing between locally constant, locally linear, and locally nonlinear behavior in each directional derivative. 
```

Comment:
```suggestion
distinguishing between locally constant, locally linear, and locally nonlinear behavior. 
```

Reply:
I want to hint at the fact that 3 classes are not enough. There are 13 types of different binary operators:
https://github.com/adrhill/SparseConnectivityTracer.jl/blob/e6d357b5e0b23d72ab3430e359831eafeb005d73/src/operators.jl#L389-L410

Reply:
I thought you meant "each directional derivative" as "each HVP". I suggested another sentence to clarify this